build/utils/npm.go

```diff
@@ -9,189 +9,284 @@ import (
 		}
-		depVersion := string(ver)
-		depId := depName + ":" + depVersion
-		if err == nil {
-			appendDependency(dependencies, depId, scope, pathToRoot)
+		if npmLsDependency.Version == "" {
+			if npmLsDependency.Missing || npmLsDependency.Problems != nil {
+				// Skip missing peer dependency.
+				log.Debug(fmt.Sprintf("%s is missing, this may be the result of an peer dependency.", key))
+				return nil
+			}
+			return errors.New("failed to parse '" + string(value) + "' from npm ls output.")
 		}
-		transitive, _, _, err := jsonparser.Get(data, depName, "dependencies")
+		if err := appendDependency(dependencies, npmLsDependency, cacache, pathToRoot, log); err != nil {
+			return err
+		}
+		transitive, _, _, err := jsonparser.Get(value, "dependencies")
 		if err != nil && err.Error() != "Key path not found" {
 			return err
 		}
 		if len(transitive) > 0 {
-			if err := parseDependencies(transitive, scope, append([]string{depId}, pathToRoot...), dependencies, log); err != nil {
+			if err := parseDependencies(transitive, append([]string{npmLsDependency.id()}, pathToRoot...), dependencies, cacache, parseFunc, log); err != nil {
 				return err
 			}
 		}
 		return nil
 	})
 }
 
-func appendDependency(dependencies *map[string]*entities.Dependency, depId, scope string, pathToRoot []string) {
-	if (*dependencies)[depId] == nil {
-		(*dependencies)[depId] = &entities.Dependency{Id: depId, Scopes: []string{scope}}
-	} else if !scopeAlreadyExists(scope, (*dependencies)[depId].Scopes) {
-		(*dependencies)[depId].Scopes = append((*dependencies)[depId].Scopes, scope)
+func parseNpmLsDependencyFunc(npmVersion *version.Version) func(data []byte) (*npmLsDependency, error) {
+	// If npm older than v7, use legacy struct for npm ls output.
+	if npmVersion.Compare("7.0.0") > 0 {
+		return legacyNpmLsDependencyParser
+	}
+	return npmLsDependencyParser
+}
+
+func legacyNpmLsDependencyParser(data []byte) (*npmLsDependency, error) {
+	legacyNpmLsDependency := new(legacyNpmLsDependency)
+	err := json.Unmarshal(data, &legacyNpmLsDependency)
+	if err != nil {
+		return nil, err
 	}
-	(*dependencies)[depId].RequestedBy = append((*dependencies)[depId].RequestedBy, pathToRoot)
+	return legacyNpmLsDependency.toNpmLsDependency(), nil
 }
 
-func scopeAlreadyExists(scope string, existingScopes []string) bool {
-	for _, existingScope := range existingScopes {
-		if existingScope == scope {
-			return true
+func npmLsDependencyParser(data []byte) (*npmLsDependency, error) {
+	npmLsDependency := new(npmLsDependency)
+	return npmLsDependency, json.Unmarshal(data, &npmLsDependency)
+}
+
+func appendDependency(dependencies map[string]*prototypeNode, dep *npmLsDependency, cacache *cacache, pathToRoot []string, log utils.Log) (err error) {
```

Comment:
Looks like error is always nil
```suggestion
func appendDependency(dependencies map[string]*prototypeNode, dep *npmLsDependency, cacache *cacache, pathToRoot []string, log utils.Log) {
```